numpy/random/mtrand.pyx

```diff
@@ -3954,10 +3993,11 @@ cdef class RandomState:
 
         if np.any(np.less(np.add(ongood, onbad), onsample)):
             raise ValueError("ngood + nbad < nsample")
-        # Convert to int64, if necessary, to use int64 infrastructure
-        ongood = ongood.astype(np.int64)
-        onbad = onbad.astype(np.int64)
-        onsample = onsample.astype(np.int64)
+        if np.any(onsample > LONG_MAX):
+            raise ValueError(
+                    "`nsample` is out of bounds for long, consider using "
+                    "the new generator API for 64bit integers.")
+
```

Comment:
There is a constraint check for `onsample` below in `discrete_broadcast_iii`. Could avoid an extra check by rolling that into the existing check.